topology/dynamic.go

```diff
@@ -0,0 +1,208 @@
+		return nil, errNoConfigServiceClient
+	}
+	services := options.GetConfigServiceClient().Services()
+	if services == nil {
+		return nil, errInvalidConfigServiceClient
+	}
+	w, err := services.WatchInstances(options.GetService(), options.GetQueryOptions())
+	if err != nil {
+		return nil, err
+	}
+	s := xwatch.NewSource(newServiceTopologyInput(w, options.GetRetryOptions()), options.GetInstrumentOptions().GetLogger())
+	// wait on source to be ready
+	err = waitOnInit(s, options.GetInitTimeout())
+	if err != nil {
+		return nil, err
+	}
+	dt := &dynamicTopology{w: w, s: s, logger: options.GetInstrumentOptions().GetLogger()}
+	return dt, nil
+}
+
+func (t *dynamicTopology) Get() Map {
+	return t.s.Get().(Map)
+}
+
+func (t *dynamicTopology) GetAndSubscribe() (Map, xwatch.Watch, error) {
+	m, w, err := t.s.Watch()
+	return m.(Map), w, err
+}
+
+func (t *dynamicTopology) Close() {
+	t.Lock()
+	defer t.Unlock()
+	if t.closed {
+		return
+	}
+
+	t.closed = true
+
+	t.w.Close()
+	t.s.Close()
+}
+
+func waitOnInit(t xwatch.Source, d time.Duration) error {
+	if d <= 0 {
+		return nil
+	}
+	select {
+	case <-t.WaitInit():
+		return nil
+	case <-time.After(d):
+		return errTimeOutInitializing
+	}
+}
+
+// serviceTopologyInput implements xwatch.SourceInput with retry
+type serviceTopologyInput struct {
+	w       xwatch.Watch
+	retrier xretry.Retrier
+	closed  bool
+	// NB(chaowang) the watch may be initiated and not get any updates on its channel any time soon,
+	// so we should try whatever value the watch has as first poll to avoid source init time out.
+	// If the value the watch has is invalid, the source is still not initiated
+	initPoll bool
+}
+
+func newServiceTopologyInput(w xwatch.Watch, opt xretry.Options) xwatch.SourceInput {
+	return &serviceTopologyInput{w: w, retrier: xretry.NewRetrier(opt), initPoll: true}
+}
+
+func (s *serviceTopologyInput) Poll() (interface{}, error) {
+	var m Map
+
+	if attemptErr := s.retrier.Attempt(
+		func() error {
+			var err error
+			if !s.initPoll {
+				// for normal attempts (from 2nd attempt on), block on the notification channel
+				if _, ok := <-s.w.C(); !ok {
+					s.closed = true
+					return xwatch.ErrSourceClosed
+				}
+			} else {
+				s.initPoll = false
+			}
+
+			if m, err = getMap(s.w.Get()); err != nil {
+				return err
+			}
+			return nil
+		}); attemptErr != nil {
+		return nil, attemptErr
+	}
+
+	return m, nil
+}
+
+func getMap(data interface{}) (Map, error) {
+	si, ok := data.([]services.ServiceInstance)
+	if !ok {
+		return nil, errInvalidTopology
+	}
+	to, err := buildTopologyOptionsFromServiceInstances(si)
+	if err != nil {
+		return nil, err
+	}
+	if err := to.Validate(); err != nil {
+		return nil, err
+	}
+	return newStaticMap(to), nil
+}
+
+func buildTopologyOptionsFromServiceInstances(instances []services.ServiceInstance) (MapOptions, error) {
+	allShards, r := getUniqueShardsAndReplicas(instances)
+
+	fn := sharding.DefaultHashFn(uint32(len(allShards)))
```

Comment:
I don't think we should specify the hashing function in concrete here.  We should probably pass this all the way in the top to new dynamic topology constructor.